schema-layer/data-structures/generics.md

```diff
@@ -0,0 +1,79 @@
+# Specification: IPLD Generics
+
+**Status: Prescriptive - Exploratory**
+
+Organizing IPLD data into usable, efficient, complex data structures spanning many blocks aimed for use by end-user applications.
+
+This document will re-use some terms found in the [IPLD data model](/IPLD-Data-Model-v1.md). 
+
+IPLD Generics offer codec agnostic programming interfaces for all common operations users can currently accomplish on Data-Model [“Kinds”]().
+
+Contents:
+
+  * [Motivation](#motivation)
+  * [Requirements](#requirements)
+  * [Implementations](#implementations)
+
+## Motivation
+
+Even before the IPLD Data-Model was formally specified developers were creating multi-block data-structures with similar semantics to single-block primitives. The most illustrative example of this is the `dag-pb` HAMT implementation used by IPFS for large directories.
+
+These early implementations of multi-block data structures exposed several problems.
+
+  * They are lacking self-description. A consumer of a graph containing these structures would have to have logic on top of IPLD and vary the way it performs operations on there data structures.
+  * Implementations of these data structures cannot perform operations on each other. In other words, multi-block data structures have a hard time building on top of each other.
+
+Since there wasn’t a standardized way to describe these data structures we couldn’t build libraries for paths and selectors that seamlessly supported them.
+
+As we started designing this system several other requirements surfaced.
+
+* Transparent encryption envelopes on the read and write side.
+* Advanced `Link` types that can support some form of mutability and link to paths within other data structures.
+* Flexible multi-block binary types.
+
+## Requirements
+
+IPLD Generics cannot be implemented without:
+
+  * The IPLD Data-Model. There are codec agnostic but do require that the full data model be implemented by the codec.
+  * Reserved `_type` property.
+
+While the `_type` property is reserved in every map regardless of where it appears in a block, IPLD Generics **MUST** include the value of `_type` in the root property of a discrete block. Note: you can still embed multiple IPLD Generics in a single block using [inline blocks]().
```

Comment:
The motivation for this section is a bit complicated, but it comes down to making the implementation easier to reason about, consistent, and performant. After a block is decoded and used by the engine you completely lose track of whether or not it is at the root, so you need the property reserved at every layer. 

However, the individual generic implementations may traverse into the node data on their own and could easily skip over a `_type` property and there’s no good way to keep them from doing so unless we don’t send them the decoded node data and require them to call back for read operations for **everything** which would be prohibitively expensive. So, the only place we can **ensure** that decoded node data is wrapped in the generic implementation is at the root of the block because that’s something the generic implementations aren’t even allowed to do and have to call back into the engine to accomplish.

Reply:
Might be good to include some of this in that paragraph.

Although I'm not getting the nuance here that explains the contradiction between: "you lose track of whether or not it is at the root" and "the only place we can ensure that decoded node data is wrapped in the generic implementation is at the root of the block". It seems like the implementation difficulties are dictating that the root is treated as a special case anyway.

Reply:
`[inline blocks]()` needs a target or to not be a link (maybe add a TODO in brackets?)

Reply:
To clarify, the *engine* loses track of whether the “node” is at the root of a block or not. This *would* enable generics in any Map except for the fact that all of the decoded block data is passed into the generic nodes, which means they are free to work through several maps and we can’t guarantee they are checking for `_type` unless we force them to call back into the engine for a read anywhere in the block data :/

Reply:
this section has been removed and replaced with a section about definitions and the open problem of fat pointers.